### PR TITLE
refactor(cardinal): move iterators into it's own package.

### PR DESCRIPTION
**Pull request by pyrofolium**
Closes: WORLD-725

iterators were living in storage but had nothing to do with "storage". There's also a Mock in here that I'll modify in another PR. 

## Summary by CodeRabbit

- **Refactor**
	- Transitioned from using the `storage` package to the `iterators` package for enhanced entity and component management within the system.
- **Bug Fixes**
	- Improved error handling related to component mismatches and entity identification, ensuring more accurate error reporting.
- **New Features**
	- Enhanced iteration over archetypes in the ECS for more efficient processing.

### CODE REVIEW

File: cardinal/ecs/ecb/read_only.go [hunk 1]
```diff
@@ -144,8 +144,8 @@ func (r *readOnlyManager) GetEntitiesForArchID(archID archetype.ID) ([]entity.ID
 	return ids, nil
 }
 
-func (r *readOnlyManager) SearchFrom(filter filter.ComponentFilter, start int) *storage.ArchetypeIterator {
-	itr := &storage.ArchetypeIterator{}
+func (r *readOnlyManager) SearchFrom(filter filter.ComponentFilter, start int) *iterators.ArchetypeIterator {
+	itr := &iterators.ArchetypeIterator{}
```
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/ecs/ecb/read_only.go:148]: This is wrong.

File: cardinal/ecs/search.go [hunk 2]
```diff
@@ -93,7 +93,7 @@ func (q *Search) First(eCtx EngineContext) (id entity.ID, err error) {
 			return entities[0], nil
 		}
 	}
-	return storage.BadID, eris.Wrap(err, "")
+	return iterators.BadID, eris.Wrap(err, "")
```
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/ecs/search.go:96]: Wrong.

File: cardinal/ecs/search.go [hunk 3]
```diff
@@ -79,9 +79,9 @@ func (q *Search) First(eCtx EngineContext) (id entity.ID, err error) {
 	namespace := eCtx.GetEngine().Namespace()
 	reader := eCtx.StoreReader()
 	result := q.evaluateSearch(namespace, reader)
-	iter := storage.NewEntityIterator(0, reader, result)
+	iter := iterators.NewEntityIterator(0, reader, result)
```
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/ecs/search.go:82]: Wrong

**Comment by pyrofolium** [comment]
> ## Walkthrough
> The codebase has undergone a significant refactor, where the `iterators` functionality has been extracted from the `storage` package and encapsulated within its own dedicated package. This change affects multiple files across the codebase, involving updates to imports, function calls, type declarations, and error handling mechanisms. The refactor aims to enhance modularity and maintainability by segregating a distinct aspect of the system (iterators) into a standalone package.
> 
> ## Changes
> Files	Change Summary
> `cardinal/ecs/ecb/ecb.go`, `cardinal/ecs/ecb/redis.go`	Imported `iterators` package, removed `storage` import, updated error references from `storage` to `iterators`.
> `.../ecb_test.go`, `.../recovery_test.go`, `.../ecs_test.go`, `.../engine.go`, `.../filter_test.go`, `.../search.go`, `.../state_test.go`, `.../tick_test.go`, `.../errors_test.go`, `.../component_test.go`, `.../world.go`	Added `iterators` imports, removed `storage` imports, replaced error constants from `storage` to `iterators`.
> `cardinal/ecs/ecb/read_only.go`, `cardinal/ecs/store/iface.go`	Replaced `storage.ArchetypeIterator` with `iterators.ArchetypeIterator`.
> `cardinal/ecs/iterators/constants.go`, `cardinal/ecs/iterators/index.go`, `cardinal/ecs/iterators/iterator.go`	Renamed package from "storage" to "iterators".
> ## Assessment against linked issues
> Objective	Addressed	Explanation
> Separate iterators from the Storage component (WORLD-725).	✅	
> Create a new package specifically for iterators (WORLD-725).	✅	
> Ensure that the codebase functions properly after the iterators are moved (WORLD-725).	❓	Without specific test results or manual testing instructions, it's unclear if functionality is unaffected post-refactor.
> Thank you for using CodeRabbit. We offer it for free to the OSS community and would appreciate your support in helping us grow. If you find it useful, would you consider giving us a shout-out on your favorite social media?
> 
> Share
> Tips

It's tested. The existing unit tests still run.

File: cardinal/ecs/search.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/ecs/search.go:82]: If err is not defined in the scope how is it even compiling? How is it passing unit tests?

**Review by technicallyty — APPROVED** [review]

**Pull request merged**